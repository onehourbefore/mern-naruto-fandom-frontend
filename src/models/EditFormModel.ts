

export enum TasksEnum {
    CREATE= 'create',
    UPDATE = 'update',
    IDLE = 'idle'
}

export interface EditFormState {
    postID: string
    task: TasksEnum
}