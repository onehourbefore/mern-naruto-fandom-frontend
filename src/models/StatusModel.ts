export enum StatusModel {
    LOADING = 'loading',
    ERROR = 'error',
    SUCCESS = 'success',
    IDLE = 'idle'
}