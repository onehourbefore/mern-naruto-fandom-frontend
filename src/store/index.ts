import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { useDispatch, TypedUseSelectorHook, useSelector } from 'react-redux';

import posts from './slices/postsSlice';
import filter from './slices/filterSlice';
import onePost from './slices/onePostSlice';
import registration from './slices/registrationSlice';
import authorization from './slices/authorizationSlice';
import experts from './slices/expertSlice';
import comments from './slices/commentSlice';
import editForm from './slices/editFormSlice';

export const rootReducer = combineReducers ({
    posts,
    filter,
    onePost,
    registration,
    authorization,
    experts,
    comments,
    editForm,
});

export const store = configureStore ({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware (),
    devTools: process.env.NODE_ENV !== 'production'
});

export type RootState = ReturnType <typeof store.getState>;
export const useAppDispatch: () => typeof store.dispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;