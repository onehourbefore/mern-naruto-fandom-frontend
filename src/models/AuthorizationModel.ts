import { StatusModel } from "./StatusModel"

export type UserDataType = {
    id: string,
    email: string,
    name: string,
    role: string,
    avatar: string,
    created: number,
    liked: string [] | []
}

export type ResponseType = {
    accessToken: string,
    refreshToken: string,
    user: UserDataType
}

export interface AuthorizationState {
    user: UserDataType,
    status: StatusModel,
    deleteProfileStatus: StatusModel
}