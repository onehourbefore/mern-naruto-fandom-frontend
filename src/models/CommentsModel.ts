import { StatusModel } from "./StatusModel"

export type CommentType = {
    _id: string,
    postID: string,
    date: string,
    name: string,
    email: string,
    body: string,
    avatar: string,
}

export interface CommentsState {
    comments: [] | CommentType [],
    status: StatusModel,
}

export type CommentProps = {
    data: {
        _id: string,
        name: string,
        email: string,
        body: string,
        avatar: string,
        date: string,
    },
    postID: string | undefined,
    commentIndex: number,
}

export type CommentFormProps = {
    setVisible: (arg: boolean) => void,
}

export type CommentFromSidebarProps = {
    comment: CommentType,
}