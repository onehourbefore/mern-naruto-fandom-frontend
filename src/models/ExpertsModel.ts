import { StatusModel } from "./StatusModel"

export type ExpertType = {
    avatar: string,
    email: string,
    name: string,
    created: number
}


export interface ExpertsState {
    experts: [] | ExpertType [],
    status: StatusModel
}