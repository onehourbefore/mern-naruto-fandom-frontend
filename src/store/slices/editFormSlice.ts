import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { EditFormState, TasksEnum } from '../../models/EditFormModel';

const initialState: EditFormState = {
    postID: '',
    task: TasksEnum.IDLE,
}

export const editFormSlice = createSlice({
    name: 'modal',
    initialState,
    reducers: {
        setEditForm: (state, action: PayloadAction<EditFormState>) => {
            const { postID, task } = action.payload;
            state.postID = postID;
            state.task = task;
        }
    }
})

const { actions, reducer } = editFormSlice;
export default reducer;
export const { setEditForm } = actions;