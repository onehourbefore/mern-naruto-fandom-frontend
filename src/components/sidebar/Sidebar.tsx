import React from 'react';
import { useNavigate } from 'react-router-dom';
import { useAppDispatch, useAppSelector } from '../../store';

import { StatusModel } from '../../models/StatusModel';
import { setEditForm } from '../../store/slices/editFormSlice';
import { TasksEnum } from '../../models/EditFormModel';
import { getAuthorization } from '../../selectors';

import SearchComponent from './search/SearchComponent';
import CloudTags from './cloudTags/CloudTags';
import LastCommentsComponent from './lastComments/LastCommentsComponent';

import cl from './Sidebar.module.scss';


const Sidebar: React.FC = (): JSX.Element => {
    const navigate = useNavigate();
    const dispatch = useAppDispatch();
    const { status: authStatus, user } = useAppSelector(getAuthorization);

    function handleSetEditForm() {
        dispatch (setEditForm ({ postID: '', task: TasksEnum.CREATE }));
        navigate ('/editform');
    }

    return (
        <div className={cl.root}>
            <div>
                <SearchComponent />
                {user.role === 'admin' && 
                <button
                    data-testid='createPostBtn-test'
                    className={cl.root__createPostBtn}
                    onClick={handleSetEditForm}
                    >
                        <div>СОЗДАТЬ</div>
                        <div>НОВЫЙ ПОСТ</div>
                </button>}
            </div>
            <CloudTags />

            {authStatus === StatusModel.SUCCESS
                && <LastCommentsComponent />}
        </div>
    )
};

export default Sidebar;