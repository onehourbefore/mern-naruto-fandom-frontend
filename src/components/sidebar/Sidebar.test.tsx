import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { StatusModel } from '../../models/StatusModel';
import { TasksEnum } from '../../models/EditFormModel';
import * as reduxHooks from '../../store';
import * as actions from '../../store/slices/editFormSlice';
import Sidebar from './Sidebar';


const mockedUseAppDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');
const mockedUseAppSelector = jest.spyOn(reduxHooks, 'useAppSelector');
const mockedSetEditFormAction = jest.spyOn(actions, 'setEditForm');

const returnState = {
    list: [],
    tags: ['one', 'two', 'three'],
    status: StatusModel.IDLE,
    user: { role: '' },
    searchStatus: StatusModel.IDLE,
    posts: [{}, {}, {}],
    fetchStatus: StatusModel.IDLE,
    comments: [],
    task: TasksEnum.CREATE,
};

describe('Sidebar', () => {
    it('no auth', () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue(returnState);

        const component = render(
            <MemoryRouter>
                <Sidebar />
            </MemoryRouter>
        );
        expect(screen.getByPlaceholderText(/поиск по блогу.../i)).toBeInTheDocument();
        expect(screen.queryByText(/создать/i)).toBeNull();
        expect(screen.queryByText(/новый пост/i)).toBeNull();
        expect(screen.getByText('one')).toBeInTheDocument();
        expect(screen.getByText('two')).toBeInTheDocument();
        expect(screen.getByText('three')).toBeInTheDocument();
        expect(screen.queryByText(/последние комментарии/i)).toBeNull();
        expect(component).toMatchSnapshot();
    });

    it('auth: { role: admin }', () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue({
            ...returnState,
            status: StatusModel.SUCCESS,
            user: { role: 'admin' },
        });

        const component = render(
            <MemoryRouter>
                <Sidebar />
            </MemoryRouter>
        );
        expect(screen.getByPlaceholderText(/поиск по блогу.../i)).toBeInTheDocument();
        expect(screen.getByText(/создать/i)).toBeInTheDocument();
        expect(screen.getByText(/новый пост/i)).toBeInTheDocument();
        expect(screen.getByText('one')).toBeInTheDocument();
        expect(screen.getByText('two')).toBeInTheDocument();
        expect(screen.getByText('three')).toBeInTheDocument();
        expect(screen.queryByText(/последние комментарии/i)).toBeNull();

        userEvent.click(screen.getByTestId('createPostBtn-test'));
        expect(mockedSetEditFormAction).toHaveBeenCalledTimes(1);
        expect(mockedSetEditFormAction).toHaveBeenCalledWith({
            postID: '',
            task: TasksEnum.CREATE,
        });
        expect(component).toMatchSnapshot();
    });
});