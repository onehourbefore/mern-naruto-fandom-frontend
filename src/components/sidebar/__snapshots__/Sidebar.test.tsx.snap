// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar auth: { role: admin } 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <div>
          <div
            class="root"
          >
            <div
              class="root__search"
            >
              <div
                class="root__searchWrapper"
              >
                <img
                  alt="Поиск"
                  class="root__search_icon"
                  src="search.png"
                />
                <input
                  class="root__search_input"
                  id="searchInput"
                  placeholder="Поиск по блогу..."
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <button
            class="root__createPostBtn"
            data-testid="createPostBtn-test"
          >
            <div>
              СОЗДАТЬ
            </div>
            <div>
              НОВЫЙ ПОСТ
            </div>
          </button>
        </div>
        <div
          class="root"
        >
          <h3
            class="root__title"
          >
            ОБЛАКО ТЕГОВ
          </h3>
          <div
            class="root__body"
          >
            <span
              class="root__tag"
            >
              one
            </span>
            <span
              class="root__tag"
            >
              two
            </span>
            <span
              class="root__tag"
            >
              three
            </span>
          </div>
        </div>
        <div
          class="root"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <div>
        <div
          class="root"
        >
          <div
            class="root__search"
          >
            <div
              class="root__searchWrapper"
            >
              <img
                alt="Поиск"
                class="root__search_icon"
                src="search.png"
              />
              <input
                class="root__search_input"
                id="searchInput"
                placeholder="Поиск по блогу..."
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <button
          class="root__createPostBtn"
          data-testid="createPostBtn-test"
        >
          <div>
            СОЗДАТЬ
          </div>
          <div>
            НОВЫЙ ПОСТ
          </div>
        </button>
      </div>
      <div
        class="root"
      >
        <h3
          class="root__title"
        >
          ОБЛАКО ТЕГОВ
        </h3>
        <div
          class="root__body"
        >
          <span
            class="root__tag"
          >
            one
          </span>
          <span
            class="root__tag"
          >
            two
          </span>
          <span
            class="root__tag"
          >
            three
          </span>
        </div>
      </div>
      <div
        class="root"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Sidebar no auth 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <div>
          <div
            class="root"
          >
            <div
              class="root__search"
            >
              <div
                class="root__searchWrapper"
              >
                <img
                  alt="Поиск"
                  class="root__search_icon"
                  src="search.png"
                />
                <input
                  class="root__search_input"
                  id="searchInput"
                  placeholder="Поиск по блогу..."
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="root"
        >
          <h3
            class="root__title"
          >
            ОБЛАКО ТЕГОВ
          </h3>
          <div
            class="root__body"
          >
            <span
              class="root__tag"
            >
              one
            </span>
            <span
              class="root__tag"
            >
              two
            </span>
            <span
              class="root__tag"
            >
              three
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <div>
        <div
          class="root"
        >
          <div
            class="root__search"
          >
            <div
              class="root__searchWrapper"
            >
              <img
                alt="Поиск"
                class="root__search_icon"
                src="search.png"
              />
              <input
                class="root__search_input"
                id="searchInput"
                placeholder="Поиск по блогу..."
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="root"
      >
        <h3
          class="root__title"
        >
          ОБЛАКО ТЕГОВ
        </h3>
        <div
          class="root__body"
        >
          <span
            class="root__tag"
          >
            one
          </span>
          <span
            class="root__tag"
          >
            two
          </span>
          <span
            class="root__tag"
          >
            three
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
